pallets/automation-price/src/lib.rs

```diff
@@ -0,0 +1,700 @@
+	), Task<T>>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn get_task_queue)]
+	pub type TaskQueue<T: Config> = StorageValue<_, Vec<(Vec<u8>, T::Hash)>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn is_shutdown)]
+	pub type Shutdown<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn get_asset_metadata)]
+	pub type AssetMetadata<T: Config> = StorageMap<_, Twox64Concat, Vec<u8>, AssetMetadatum<T>>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The provided_id cannot be empty
+		EmptyProvidedId,
+		/// Time must end in a whole hour.
+		InvalidTime,
+		/// Duplicate task
+		DuplicateTask,
+		/// Non existent asset
+		AssetNotSupported,
+		/// Asset already supported
+		AssetAlreadySupported,
+		/// Asset cannot be updated by this account
+		InvalidAssetSudo,
+		/// Asset must be in triggerable range.
+		AssetNotInTriggerableRange,
+		/// Block Time not set
+		BlockTimeNotSet,
+		/// Invalid Expiration Window for new asset
+		InvalidAssetExpirationWindow,
+		/// Maximum tasks reached for the slot
+		MaxTasksReached,
+		/// Failed to insert task
+		TaskInsertionFailure,
+		/// Insufficient Balance
+		InsufficientBalance,
+		/// Restrictions on Liquidity in Account
+		LiquidityRestrictions,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Schedule task success.
+		TaskScheduled {
+			who: AccountOf<T>,
+			task_id: T::Hash,
+		},
+		Notify {
+			message: Vec<u8>,
+		},
+		TaskNotFound {
+			task_id: T::Hash,
+		},
+		AssetCreated {
+			asset: Vec<u8>,
+		},
+		AssetUpdated {
+			asset: Vec<u8>,
+		},
+		AssetDeleted {
+			asset: Vec<u8>,
+		},
+		AssetPeriodReset {
+			asset: Vec<u8>,
+		},
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_: T::BlockNumber) -> Weight {
+			if Self::is_shutdown() == true {
+				return T::DbWeight::get().reads(1 as Weight)
+			}
+
+			let max_weight: Weight =
+				T::MaxWeightPercentage::get().mul_floor(T::MaxBlockWeight::get());
+			Self::trigger_tasks(max_weight)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Schedule a task to fire an event with a custom message.
+		///
+		/// Before the task can be scheduled the task must past validation checks.
+		///
+		/// # Parameters
+		/// * `provided_id`: An id provided by the user. This id must be unique for the user.
+		/// * `asset`: asset type
+		/// * `direction`: direction of trigger movement
+		/// * `trigger_percentage`: what percentage task should be triggered at 
+		///
+		/// # Errors
+		#[pallet::weight(<T as Config>::WeightInfo::schedule_swap_task_extrinsic())]
+		#[transactional]
+		pub fn schedule_swap_task(
+			origin: OriginFor<T>,
+			provided_id: Vec<u8>,
+			asset: Vec<u8>,
+			direction: u8,
+			trigger_percentage: u128,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let fee = <BalanceOf<T>>::saturated_from(1_000_000_000_000u64);
```

Comment:
i figured i'd make it expensive, lemme know if it's too expensive (it's like 1-1.5 TUR)